app/code/Magento/InventoryCatalog/Model/ResourceModel/BulkInventoryTransfer.php

```diff
@@ -0,0 +1,186 @@
+        ], [
+            SourceItemInterface::SKU . ' = ?' => $sku,
+            $sourceCodeCondition,
+        ]);
+
+        return $quantityExtracted;
+    }
+
+    /**
+     * Transfer quantity
+     * @param string $sku
+     * @param string $destinationSource
+     * @param float $quantity
+     */
+    private function transferQuantity(string $sku, string $destinationSource, float $quantity)
+    {
+        $connection = $this->resourceConnection->getConnection();
+        $tableName = $this->resourceConnection->getTableName(SourceItem::TABLE_NAME_SOURCE_ITEM);
+
+        $query = $connection->select()->from($tableName, new Expression('SUM(' . SourceItemInterface::QUANTITY . ')'))
+            ->where(SourceItemInterface::SOURCE_CODE . ' = ?', $destinationSource)
+            ->where(SourceItemInterface::SKU . ' = ?', $sku);
+
+        $qtyDestinationSource = $connection->fetchOne($query);
+        if ($qtyDestinationSource === null) {
+            $connection->insert($tableName, [
+                SourceItemInterface::SOURCE_CODE => $destinationSource,
+                SourceItemInterface::SKU => $sku,
+                SourceItemInterface::QUANTITY => $quantity,
+                SourceItemInterface::STATUS => $quantity > 0 ?
+                    SourceItemInterface::STATUS_IN_STOCK :
+                    SourceItemInterface::STATUS_OUT_OF_STOCK
+            ]);
+        } else {
+            $finalQty = $quantity + (float) $qtyDestinationSource;
+
+            $connection->update($tableName, [
+                SourceItemInterface::QUANTITY => $finalQty,
+                SourceItemInterface::STATUS => $finalQty > 0 ?
+                    SourceItemInterface::STATUS_IN_STOCK :
+                    SourceItemInterface::STATUS_OUT_OF_STOCK
+            ], [
+                SourceItemInterface::SOURCE_CODE . '=?' => $destinationSource,
+                SourceItemInterface::SKU . '=?' => $sku,
+            ]);
+        }
+    }
+
+    /**
+     * Assign sources to products
+     * @param array $skus
+     * @param string $destinationSource
+     * @param bool $defaultSourceOnly
+     * @return void
+     */
+    public function execute(array $skus, string $destinationSource, bool $defaultSourceOnly = false): void
+    {
+        $types = $this->getProductTypesBySkus->execute($skus);
+        $connection = $this->resourceConnection->getConnection();
+
+        foreach ($types as $sku => $type) {
+            if ($this->isSourceItemManagementAllowedForProductType->execute($type)) {
+                $connection->beginTransaction();
+
+                $sourceItemsIds = [];
+                $totalQty = $this->extractSourcesQuantity(
+                    $sku,
+                    $destinationSource,
+                    $sourceItemsIds,
+                    $defaultSourceOnly
+                );
+
+                $this->transferQuantity($sku, $destinationSource, $totalQty);
```

Comment:
such kind of long-running operations is a perfect candidate for 
`Magento\AsynchronousOperations` and `\Magento\Framework\Bulk\BulkManagementInterface`

to schedule an asynchronous job